Feature/blk 68 fix gas under metering in permit precompile

sfffaaa opened the pull request.

Commit:
Add more permit tests

Commit:
Add comprehensive EIP-712 permit testing to bridge_balance_erc20_test

  - Implement permit signature generation using EIP-712 structured data
  - Add dual-contract pattern for permit and ERC20 function separation
  - Test valid permit workflow with allowance verification
  - Add security tests for expired deadlines and invalid signatures
  - Handle contract reverts properly during gas estimation
  - Use ETH/balance-erc20-permit/abi for permit-specific functions

Commit:
fix flake8

Commit:
Fix the review suggestion

neutrinoks commented on tests/bridge_balance_erc20_test.py:
```diff
@@ -241,3 +325,173 @@ def test_transfer_to_account_id(self):
+        self.assertEqual(allowance, permit_value, f'Allowance not set correctly: {allowance} != {permit_value}')
+
+        # Verify nonce was incremented
+        new_nonce = permit_contract.functions.nonces(self._eth_kp_src['eth']).call()
```
:raised_hands:

neutrinoks commented on tests/bridge_balance_erc20_test.py:
```diff
@@ -241,3 +325,173 @@ def test_transfer_to_account_id(self):
+        # Execute permit
+        permit_receipt = self.evm_permit(
+            permit_contract,
+            self._eth_kp_src['kp'],
```
Shouldn't we test to see if someone else can hand in that permit? Or did I misunderstand the code?

neutrinoks reviewed (commented):
Looks good. One question left that seems relevant to me in case of using the permit function.

Commit:
Add comprehensive ERC-20 permit testing with third-party submission

  - Add test_permit_third_party_submission to verify EIP-2612 permit relaying functionality
  - Refactor evm_permit function to support different transaction submitters
  - Remove unnecessary try-catch blocks for permit availability checks
  - Add helper methods _fund_accounts_for_permit and _verify_permit_success for better maintainability
  - Maintain detailed f-string error messages for debugging
  - Test covers owner signs permit, third party submits transaction, spender gets allowance

neutrinoks reviewed (approved):
:rocket:
